### PR TITLE
Add pip_data_exclude to pip_repository

### PR DESCRIPTION
**Pull request by gergelyfabian**
This can be used to extend the list of exclusions for Bazel packages
generated from pip wheels.

### CODE REVIEW

**Comment by gergelyfabian** [comment]
This can be used as a general solution, e.g. to exclude all files that have folder names that include spaces (that would break Bazel build).
Also useful for ad-hoc workarounds for users (e.g. when working with remote caching).

**Comment by gergelyfabian** [comment]
@dillon-giacoppo Anything else necessary here after #42 is merged?

File: defs.bzl
Comment (dillon-giacoppo) [inline defs.bzl]: Same thought as #41, can you confirm that wrapping in quotes is sufficient and we don't need full shell quoting like what you would get from `https://docs.python.org/3/library/shlex.html#shlex.quote`.

Comment (gergelyfabian) [reply defs.bzl]: I checked it. I've added:

```
pip_data_exclude = ["\"**/* */*\""],
```

And it ended up in the `BUILD` file as:

```
data = glob(["**/*"], exclude=["**/*.py", "**/* *", "BUILD", "WORKSPACE", "**/* */*"]),
```

Comment (gergelyfabian) [reply defs.bzl]: Sorry, I was wrong. Adding:

```pip_data_exclude = ["\"**/* */*\""],```

Makes pip fail. The Python code I added into `__init__.py` makes JSON parsing, what is missing would be proper JSON encoding in defs.bzl. I did not know how to do that in Starlark. Any hints maybe?

Comment (gergelyfabian) [reply defs.bzl]: I pushed a proposed fix for this. Now you can add even quoted strings into pip_data_exclude (I used Starlark's JSON encode capabilities).

Comment (dillon-giacoppo) [reply defs.bzl]: Nice! We should backport this to the `extra_pip_args` so that we can support spaces in arguments. WDYT?

File: extract_wheels/lib/bazel.py
Comment (dillon-giacoppo) [inline extract_wheels/lib/bazel.py:24]: Why are we excluding `"**/* *"`? IIUC this means any file with a space in the name, not sure we want that.

**Review by dillon-giacoppo — APPROVED** [review]
LGTM, after confirming `**/* *` as a default exclusion pattern.

File: defs.bzl
Comment (gergelyfabian) [reply defs.bzl]: OK. I'll try to send a PR with a backport.

File: extract_wheels/lib/bazel.py
Comment (gergelyfabian) [reply extract_wheels/lib/bazel.py:24]: There are several reasons for that :)

1. I only refactored this definition, it was also in the previous version: https://github.com/dillon-giacoppo/rules_python_external/blob/master/extract_wheels/lib/bazel.py#L32
2. Afair it originates in rules_python, files with spaces in their names are also excluded in the BUILD files there
3. If you have files in the pip package (it happens quite frequently) with a space in their name then Bazel will fail when you try to build, at least on Linux this is the behavior. Bazel does not allow for files with spaces in their names, hence we need to protect from such. In fact also directories with spaces in their name may cause issues, but with the pip_data_exclude option this can be tackled by the user (we may want to add that as a default exclude if you think).

Comment (dillon-giacoppo) [reply extract_wheels/lib/bazel.py:24]: Thanks for the detailed response, appreciate it!

**Pull request merged**
